### DISCUSSION TITLE
Monkey Patching Issue

### DISCUSSION THREAD

**Discussion by spmsh** [miguelgrinberg/Flask-SocketIO#2048] [Q&A]
We recently moved to use multiple Gunicorn processes and added a message queue (redis), and monkey patching eventlet as per the documentation, it is working well, except that now we're facing issue with some function that we run using socketio.start_background_task that seems to hang forever (and blocking all other processing) since we monkey patched.

The function in question is connecting to Azure Event Hubs to receive notification very similar to this [Consume events from an Event Hub](https://learn.microsoft.com/en-us/python/api/overview/azure/eventhub-readme?view=azure-python#consume-events-from-an-event-hub) 

We found out that if we monkey patching using `eventlet.monkey_patch(thread=False)` then the function operates back to normal and the flask app is running normally.

Any idea why monkey patching would cause our function to hang, and what are the impact of running monkey patching with thread=False.

Thanks

Note: monkey patching is done at the very beginning before all other imports, and the server is run with socketio.run()

**Comment by miguelgrinberg** [comment]
I suggest you stop using eventlet if it gives you compatibility issues. Not monkey patching may be okay in some cases, but not in others, there isn't an absolute answer.

**Comment by spmsh** [comment]
This is where I am getting confused. If I remove eventlet, does the app lose support of websockets completely? 

I am running the app on both Linux and Windows (depending the environment).

On Linux it's run through Gunicorn `gunicorn --worker-class eventlet` and in Windows with `socketio.run()` (and async_mode='eventlet`) as Gunicorn is not supported.

I was under the impression that eventlent (or gevent), are mandatory to run flask socketio. Did I not understand correctly?

What are the impact if I run flask socketio without eventlet (or gevent)? 

Thanks again for your time!

PS: how about running it with message queue. If not using eventlet, we don't need to monkey patch, is the message queue still going to work?

**Reply by miguelgrinberg** [reply]
> I was under the impression that eventlent (or gevent), are mandatory to run flask socketio. Did I not understand correctly?

WebSocket works even if you don't use eventlet or gevent.

> What are the impact if I run flask socketio without eventlet (or gevent)?

You will be using threads for concurrency instead of greenlets. Threads are heavier than greenlets, which means that you will not be able to handle as many concurrent clients as with greenlets. But you will avoid all the issues of incompatibilites between greenlets and other packages.

> how about running it with message queue. If not using eventlet, we don't need to monkey patch, is the message queue still going to work?

Monkey patching is an eventlet/gevent thing that helps make greenlets more compatible with 3rd party packages. If you don't use greenlets then you don't have any compatibility issues and there is nothing to monkey patch.
